torch/utils/_sympy/value_ranges.py

```diff
@@ -500,6 +500,57 @@ def or_(a, b):
     def and_(a, b):
         return ValueRanges.coordinatewise_increasing_map(a, b, sympy.And)
 
+    @staticmethod
+    def _bool_to_int(x):
+        if x.is_singleton():
+            return ValueRanges.wrap(sympy.Integer(1 if x.lower else 0))
+        else:
+            return ValueRanges(sympy.Integer(0), sympy.Integer(1))
+
+    @staticmethod
+    def _bit_length(n):
+        return len(bin(n).lstrip("-0b"))
+
+    @classmethod
+    def bitwise_and(cls, a, b):
+        a, b = ValueRanges.wrap(a), ValueRanges.wrap(b)
+        if a.is_bool and b.is_bool:
+            return cls.and_(a, b)
+        if a.is_bool:
+            a = cls._bool_to_int(a)
+        if b.is_bool:
+            b = cls._bool_to_int(b)
+        lower = min(a.lower, b.lower)
+        if lower < 0 and lower != -int_oo:
+            # If both lower bounds are negative, then bits start like
+            # 1...10..., so the smallest possible value is 1...101...1.
+            # Thus, we need to find the next smallest power of 2 (inclusive).
+            lower = -(1 << cls._bit_length(-lower - 1))
+        else:
+            lower = 0
+        return ValueRanges(lower, max(a.upper, b.upper))
+
+    @classmethod
+    def bitwise_or(cls, a, b):
+        a, b = ValueRanges.wrap(a), ValueRanges.wrap(b)
+        if a.is_bool and b.is_bool:
+            return cls.or_(a, b)
+        if a.is_bool:
+            a = cls._bool_to_int(a)
+        if b.is_bool:
+            b = cls._bool_to_int(b)
+        upper = max(a.upper, b.upper)
+        if upper == 0:
+            upper = 0
+        elif upper > 0 and upper != int_oo:
+            # If both upper bounds are positive, then the largest
+            # possible value is 01...1, so we need to find
+            # next largest power of 2 (exclusive), minus 1
+            upper = (1 << cls._bit_length(upper)) - 1
+        elif upper < 0:
+            upper = -1
+        return ValueRanges(min(a.lower, b.lower), upper)
```

Comment:
This bound does not seem tight. Counterexample: Suppose we have [0b101, inf] and [0b010, inf]. Then the minimum per analysis here is 0b010.  But it is impossible to attain this minimum: 0b101 | 0b010 = 0b111, and any larger value for 0b101 or 0b010 would cause the X in 0bX00 (or a higher bit) to be set.

Reply:
I have a broader question, which is do we *really* need value range analysis here? Like can we not just say that bitwise operations preserve sign and leave it at that?

Reply:
We could certainly just say that bitwise ops preserve sign - I don't completely understand the purpose of ValueRanges, so the current implementation is a best effort bound that is not guaranteed to be tight.

Reply:
ValueRanges is used to make determinations if we statically know things to true. It is designed to be a simple but powerful reasoning mechanism. For most size variables we know their range is `[2, inf]`. So we can easily conclude that e.g., `2 * s0 != 0`.